Core/GDCore/IDE/Events/UsedExtensionsFinder.cpp

```diff
@@ -0,0 +1,137 @@
+#include "UsedExtensionsFinder.h"
+
+#include "GDCore/Events/Instruction.h"
+#include "GDCore/Events/Parsers/ExpressionParser2.h"
+#include "GDCore/Extensions/Metadata/MetadataProvider.h"
+#include "GDCore/Extensions/PlatformExtension.h"
+#include "GDCore/IDE/WholeProjectRefactorer.h"
+#include "GDCore/Project/BehaviorContent.h"
+#include "GDCore/Project/Object.h"
+#include "GDCore/Project/Project.h"
+
+namespace gd {
+
+std::set<gd::String> UsedExtensionsFinder::ScanProject(gd::Project& project) {
+  UsedExtensionsFinder worker(project);
+  gd::WholeProjectRefactorer::ExposeProjectObjects(project, worker);
+  gd::WholeProjectRefactorer::ExposeProjectEvents(project, worker);
+  return worker.usedExtensions;
+};
+
+// Objects scanner
+
+void UsedExtensionsFinder::DoVisitObject(gd::Object& object) {
+  usedExtensions.insert(gd::MetadataProvider::GetExtensionAndObjectMetadata(
+                            projectPtr->GetCurrentPlatform(), object.GetType())
+                            .GetExtension()
+                            .GetName());
+};
+
+// Behaviors scanner
+
+void UsedExtensionsFinder::DoVisitBehavior(gd::BehaviorContent& behavior) {
+  usedExtensions.insert(
+      gd::MetadataProvider::GetExtensionAndBehaviorMetadata(
+          projectPtr->GetCurrentPlatform(), behavior.GetTypeName())
+          .GetExtension()
+          .GetName());
+};
+
+// Instructions scanner
+
+bool UsedExtensionsFinder::DoVisitInstruction(gd::Instruction& instruction,
+                                              bool isCondition) {
+  auto metadata =
+      isCondition
+          ? gd::MetadataProvider::GetExtensionAndConditionMetadata(
+                projectPtr->GetCurrentPlatform(), instruction.GetType())
+          : gd::MetadataProvider::GetExtensionAndActionMetadata(
+                projectPtr->GetCurrentPlatform(), instruction.GetType());
+  usedExtensions.insert(metadata.GetExtension().GetName());
+
+  for (auto expression : instruction.GetParameters()) {
+    gd::ExpressionParser2 parser(projectPtr->GetCurrentPlatform(),
+                                 GetGlobalObjectsContainer(),
+                                 GetObjectsContainer());
+    parser.ParseExpression("unknown", expression.GetPlainString())
+        ->Visit(*this);
```

Comment:
This might be a bit too direct. Iterate on the metadata parameters, so that for each parameter you have its type and can check if it's a number or a string (`ParameterMetadata::IsExpression("number", type) || ParameterMetadata::IsExpression("string", type)`) and if so, use this type for the parser (`parser.ParseExpression(type, ...)`).

Otherwise, the parameter is not an expression and it would not be useful to parse it. Passing the proper type is also important, because I guess otherwise the parsing could be slightly different at some points.

See an example in EventsRefactorer::RenameObjectInConditions (this class is a bit more complicated than it should because it's not using `ArbitraryEventsWorker`, but it's still doing its job properly for parameters)